### DISCUSSION TITLE
Meta3D目前正在内测，欢迎大家使用、贡献、交流

### DISCUSSION THREAD

**Discussion by yyc-git** [Meta3D-Technology/Meta3D#8] [Announcements]
Meta3D目前正在内测，欢迎大家使用、贡献、交流

有什么想法可以在这里交流，感谢大家~

有问题可先看 [FAQ](https://meta3d-website.4everland.app/docs/FAQ)

**Comment by tankant** [comment]
第一次在这里协作开发，希望我能如愿开发出3D养鱼的游戏来。

**Reply by yyc-git** [reply]
欢迎交流~Meta3D是做引擎和编辑器的，不应该拿来做上层的游戏，谢谢

**Comment by l1305599284** [comment]
首先Meta3D的定位是开源的Web3D低代码平台，致力于建设共享互助开放的Web3D生态，让Web3D引擎和编辑器开发轻而易举。
它有很强的扩展性和解耦性。
问题：
1.作为开发者，我想开发上层应用。就需要对扩展市场的相关依赖有所了解。
方法（1）：找到我能用的依赖，按照它提供的接口和协议规范，进行上层开发；
（这种方式成本巨大，随着扩展市场包越来越多，我需要的扩展包找不到；或者我需要多个扩展包，还要对每个包研究分析；或者我需要的扩展包又依赖了别的扩展包；）
方法（2）：自己写依赖包，进行下层开发，开发完了再进行上层开发；
（这种方式成本也较大，如果底层依赖和上层都是一个人开发的，那么可以完全脱离Meta3D平台）

2.作为使用者，不了解编程，使用起来非常困难；
（首先使用者不懂内部机制，和拼接规则。另外Meta3D不能保证任意包都是能直接运行生成结果的。）

**Reply by yyc-git** [reply]
感谢您的反馈~

>  方法（1）：找到我能用的依赖，按照它提供的接口和协议规范，进行上层开发；

是的，比如您想开发一个引擎，如果目前扩展市场中已经有了场景管理的扩展，那么你只需要直接使用它，而只实现渲染扩展。在实现渲染扩展时，该扩展依赖场景管理扩展的协议。它的协议应该给出了它能提供什么场景数据的接口，然后您在渲染扩展中直接调用协议定义的接口来获得场景数据。

**Reply by yyc-git** [reply]
>（这种方式成本巨大，随着扩展市场包越来越多，我需要的扩展包找不到；或者我需要多个扩展包，还要对每个包研究分析；或者我需要的扩展包又依赖了别的扩展包；）

如果您需要的扩展包找不到，您就要开发自己的扩展包；
如果您需要多个扩展包或者又依赖了别的扩展包，那么您只需要知道它们的协议，而不需要知道具体的实现。

**Reply by yyc-git** [reply]
> 方法（2）：自己写依赖包，进行下层开发，开发完了再进行上层开发；
（这种方式成本也较大，如果底层依赖和上层都是一个人开发的，那么可以完全脱离Meta3D平台）

您这里说的下层开发应该是指开发扩展，上层开发应该是指组装多个扩展为编辑器或者引擎吧

这里的上层开发其实都是进行可视化地拖拽，不需要写代码，所以真正的成本就只是开发您需要的扩展而已（已有的扩展则直接拿来用或者修改）

所以说其实只有下层开发，也就是开发扩展。

因为扩展是按照Meta3D的规范来写的，所以扩展只能基于Meta3D使用，所以“脱离Meta3D平台”应该是指能够将搭建好的编辑器或者引擎下载到本地来部署或者使用。
目前引擎已经能够下载到本地了。而编辑器暂时只能托管在平台上运行，以后应该会支持下载到本地部署

**Reply by yyc-git** [reply]
> 2.作为使用者，不了解编程，使用起来非常困难；
（首先使用者不懂内部机制，和拼接规则

使用者确实需要编程才能实现扩展，但是搭建编辑器和引擎却不需要编程

所以说Meta3D是低代码而不是无代码平台，对于扩展的开发还是需要编程的


> 另外Meta3D不能保证任意包都是能直接运行生成结果的。

是的，这个需要开发扩展的使用者来负责该扩展与其它的扩展一起搭建时能够运行成功，具体的测试调试流程请详见：[测试和调试扩展、贡献](https://meta3d-website.4everland.app/docs/%E6%B5%8B%E8%AF%95%E5%92%8C%E8%B0%83%E8%AF%95%E6%89%A9%E5%B1%95%E3%80%81%E8%B4%A1%E7%8C%AE)

**Reply by l1305599284** [reply]
> 感谢您的反馈~
> 
> > 方法（1）：找到我能用的依赖，按照它提供的接口和协议规范，进行上层开发；
> 
> 是的，比如您想开发一个引擎，如果目前扩展市场中已经有了场景管理的扩展，那么你只需要直接使用它，而只实现渲染扩展。在实现渲染扩展时，该扩展依赖场景管理扩展的协议。它的协议应该给出了它能提供什么场景数据的接口，然后您在渲染扩展中直接调用协议定义的接口来获得场景数据。

如果我想开发一个引擎，就需要去扩展市场找能用的依赖；甚至开始阶段我不知道我需要哪些扩展，应该从哪里上手；去寻找已有依赖的话，这个寻找的难度太大，没法很快找到一个适用的扩展；而且我不知道一个已提交的扩展能不能适合我，这就需要再去研究、查看该扩展；这块有好的方式去解决吗？

**Reply by yyc-git** [reply]
有三个方法可以找到适合的扩展：

- 查看该扩展实现的扩展协议或者扩展的description，它定义在package.json->description字段，简要描述了在做什么，如下图红框所示
![image](https://img2023.cnblogs.com/blog/419321/202302/419321-20230222152029064-1973615064.jpg)




- 查看该扩展实现的扩展协议的repo，看它的README以及定义的接口
如果该扩展协议的package.json->repoLink字段中给出了它的repo的地址，那么就可以在扩展市场的中看到它的Repo链接，点进去就可以进入，如下图所示：

![image](https://img2023.cnblogs.com/blog/419321/202302/419321-20230222151347668-807278794.jpg)


- 查看该扩展的repo，看它的README以及代码
如果该扩展的package.json->repoLink字段中给出了它的repo的地址，那么就可以在扩展市场的中看到它的Repo链接，点进去就可以进入，如下图所示：

![image](https://img2023.cnblogs.com/blog/419321/202302/419321-20230222151350271-965780561.jpg)

**Reply by l1305599284** [reply]
如果扩展市场变得很庞大，一个个看readme或者看源码的方式很不可取

**Reply by yyc-git** [reply]
后面会增加搜索功能

**Reply by yyc-git** [reply]
以后也可以考虑增加“自动推荐相关扩展”的功能，它能根据您输入的关键字或者识别您的扩展所依赖的其它扩展协议来自动为您推荐

**Reply by l1305599284** [reply]
> 以后也可以考虑增加“自动推荐相关扩展”的功能，它能根据您输入的关键字或者识别您的扩展所依赖的其它扩展协议来自动为您推荐

ok
